radix-engine/src/engine/process.rs

```diff
@@ -503,14 +507,18 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
         proofs: HashMap<ProofId, Proof>,
     ) -> Result<(), RuntimeError> {
         if self.depth == 0 {
-            assert!(proofs.is_empty());
-
+            // buckets are aggregated by worktop
             for (_, bucket) in buckets {
                 self.worktop
                     .put(bucket)
                     .map_err(RuntimeError::WorktopError)?;
             }
+            // proofs are accumulated by auth worktop
+            for (_, proof) in proofs {
```

Comment:
"A proof can go up the stack many times and down once". It will actually be the most common use case, where a proof is returned by an account component and used for system auth later. 

These proofs are explicitly returned by components. Dangling proofs (either in auth zone or not) are automatically dropped when a process exits, which is one of the TODOs.

Reply:
Ah got it